posts/2023-09-07-generating-a-rust-client-library-for-zio-http-endpoints.md

```diff
@@ -0,0 +1,717 @@
+
+What are `rawTemplateId` and `workerName`? These are so called **path codecs**, defined in a common place so they can be reused in multiple endpoints. They allow us to have dynamic parts of the request path mapped to specific types - so when we implement the endpoint (or call in a client) we don't have to pass strings, we can directly work with the business domain types, in this case `RawTemplateId` and `WorkerName`.
+
+The simplest way to define path codecs is to **transform** and existing one:
+
+```scala
+val workerName: PathCodec[WorkerName] =
+  string("worker-name").transformOrFailLeft(WorkerName.make(_).toErrorEither, _.value)
+```
+
+Here the `make` function is a **ZIO Prelude** [`Validation`](https://zio.github.io/zio-prelude/docs/functionaldatatypes/validation) which we have to convert to an `Either` for the transform function. Validations can contain more than one failures, opposed to `Either`s, which allows us to compose them in a way that we can keep multiple errors instead of immediately returning with the first failure.
+
+The `tokenSecret` is similar, but it is a `HeaderCodec` describing what type of header it is and how the value of the given header should be mapped to a specific type (a token, in this case).
+
+What is `WorkerMetadata` and how does ZIO Http know how to produce a JSON from it?
+
+It's just a simple _case class_:
+
+```scala
+final case class WorkerMetadata(
+  workerId: ComponentInstanceId,
+  accountId: AccountId,
+  args: Chunk[String],
+  env: Map[String, String],
+  status: InstanceStatus,
+  templateVersion: Int,
+  retryCount: Int
+)
+```
+
+But with an implicit **derived** **ZIO Schema**:
+
+```scala
+object WorkerMetadata {
+  implicit val schema: Schema[WorkerMetadata] = DeriveSchema.gen[WorkerMetadata]
+}
+```
+
+We will talk more about ZIO Schema below - for now all we need to know it describes the structure of Scala types, and this information can be used to serialize data into various formats, including JSON.
+
+Once we have our endpoints defined like this, we can do several things with them - they are just data describing how and endpoint looks like!
+
+### Implementing an endpoint
+
+When developing a _server_, the most important thing to do with an endpoint is to **implement** it. Implementing an endpoint looks like the following:
+
+```scala
+val getWorkerMetadataImpl =
+    getWorkerMetadata.implement {
+      Handler.fromFunctionZIO { (rawTemplateId, workerName, authTokenId) =>
+        // ... ZIO program returning with a WorkerMetadata
+      }
+    }
+```
+
+The _type_ of `getWorkerMetadataImpl` is `Route` - it is no longer just a description of how an endpoint looks like, it defines a specific HTTP route and its associated _request handler_, implemented by a ZIO effect (remember that ZIO effects are also values - we _describe_ what we need to do when a request comes in, but executing it will be the responsibility of the server implementation).
```

Comment:
```suggestion
The _type_ of `getWorkerMetadataImpl` is `Route` - it is no longer just a description of what an endpoint looks like, it defines a specific HTTP route and its associated _request handler_, implemented by a ZIO effect (remember that ZIO effects are also values - we _describe_ what we need to do when a request comes in, but executing it will be the responsibility of the server implementation).
```